{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f85365f8-cb44-4231-af93-c2a3762c0458",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import bioframe \n",
    "import pandas as pd\n",
    "import numpy as np \n",
    "import os\n",
    "import json\n",
    "from io import StringIO\n",
    "import random\n",
    "\n",
    "import pysam\n",
    "import h5py\n",
    "\n",
    "from Bio import motifs\n",
    "from Bio import pairwise2\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7ef774f8-dc36-4ddd-bb0a-e921f088e8d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-07-20 17:33:02.815563: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /spack/apps/linux-centos7-x86_64/gcc-8.3.0/python-3.9.2-uvcroioc4witkp6qf7mbebof5ix4wlb6/lib:/spack/apps/linux-centos7-x86_64/gcc-8.3.0/pmix-3.1.3-3sm6emyqaxapunh7rwbjvtaqoqe2e5z3/lib:/spack/apps/linux-centos7-x86_64/gcc-8.3.0/openmpi-4.0.2-ipm3dnvlbtxawpi4ifz7jma6jgr7mexq/lib:/spack/apps/linux-centos7-x86_64/gcc-8.3.0/openblas-0.3.8-2no6mfziiclwxb7lstxoos335gnhjpes/lib:/spack/apps/gcc/8.3.0/lib64::/home1/smaruj/software/GSL/lib:/home1/smaruj/software/HTSLIB/lib\n",
      "2022-07-20 17:33:02.815600: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "\n",
    "sys.path.insert(0, \"/home1/smaruj/akita_utils/\")\n",
    "\n",
    "# from akita_utils import *\n",
    "import akita_utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2e1b8424-5a6d-4817-be16-899ae47ead16",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_open = pysam.Fastafile(\"/project/fudenber_735/genomes/mm10/mm10.fa\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aad5b760-b6ce-47ee-8a7a-80afbf92a493",
   "metadata": {},
   "outputs": [],
   "source": [
    "nr_strong = 2\n",
    "nr_weak = 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "35658d34-3afd-4335-91a4-f5092d24d0ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29 duplicates removed for  /project/fudenber_735/tensorflow_models/akita/v2/analysis/permute_boundaries_motifs_ctcf_mm10_model7/scd.h5\n",
      "29 duplicates removed for  /project/fudenber_735/tensorflow_models/akita/v2/analysis/permute_boundaries_motifs_ctcf_mm10_model1/scd.h5\n",
      "annotating each site with boundary-wide scores\n",
      "filtering sites by overlap with rmsk\n",
      "df prepared\n"
     ]
    }
   ],
   "source": [
    "seq_coords_df = akita_utils.prepare_insertion_tsv(\n",
    "    h5_dirs = '/project/fudenber_735/tensorflow_models/akita/v2/analysis/permute_boundaries_motifs_ctcf_mm10_model*/scd.h5',\n",
    "    score_key = 'SCD',\n",
    "    pad_flank = 0, #how much flanking sequence around the sites to include\n",
    "    weak_thresh_pct = 1, # don't use sites weaker than this, might be artifacts\n",
    "    weak_num = nr_weak ,\n",
    "    strong_thresh_pct = 99, # don't use sites weaker than this, might be artifacts\n",
    "    strong_num = nr_strong ,\n",
    "    save_tsv=None, # optional filename to save a tsv\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "58575bb9-cde2-4c82-a43d-3dd51cc6e6b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>genomic_SCD</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>chr12</td>\n",
       "      <td>35192359</td>\n",
       "      <td>35192378</td>\n",
       "      <td>-</td>\n",
       "      <td>53.656250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>102781112</td>\n",
       "      <td>102781131</td>\n",
       "      <td>-</td>\n",
       "      <td>53.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>chr7</td>\n",
       "      <td>139598311</td>\n",
       "      <td>139598330</td>\n",
       "      <td>+</td>\n",
       "      <td>0.058136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>chr8</td>\n",
       "      <td>45342919</td>\n",
       "      <td>45342938</td>\n",
       "      <td>+</td>\n",
       "      <td>0.058319</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index  chrom      start        end strand  genomic_SCD\n",
       "0      0  chr12   35192359   35192378      -    53.656250\n",
       "1      1   chr8  102781112  102781131      -    53.500000\n",
       "2      2   chr7  139598311  139598330      +     0.058136\n",
       "3      3   chr8   45342919   45342938      +     0.058319"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_coords_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "79e051b0-6b01-4def-a2d0-39f6c509f4dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_orientation(seq_coords_df, \n",
    "                    nr_strong,\n",
    "                    nr_weak,\n",
    "                    mode=\"same\", \n",
    "                    orient_list=[\">>\" for i in range(len(seq_coords_df))]):\n",
    "    \n",
    "    df_len = len(seq_coords_df)\n",
    "    \n",
    "    mode_types = [\"same\", \"customize\", \"weak_strong\", \"for_each\"]\n",
    "    if mode not in mode_types:\n",
    "        raise ValueError(\"Invalid mode. Expected one of: %s\" % mode_types)\n",
    "    \n",
    "    if mode == \"same\":\n",
    "        seq_coords_df[\"orientation\"] = [\">>\" for i in range(df_len)]\n",
    "    \n",
    "    elif mode == \"customize\":\n",
    "        if len(orient_list) == df_len:\n",
    "            seq_coords_df[\"orientation\"] = orient_list\n",
    "        else:\n",
    "            raise ValueError(\"Check the length of the customized list of orientation strings. Expected length: %s\" % df_len)\n",
    "    \n",
    "    elif mode == \"weak_strong\":\n",
    "        \n",
    "        if len(orient_list) != 2:\n",
    "            raise ValueError(\"You should provide a list of two orientation strings: first of weak motifs, second - for strong\")\n",
    "        \n",
    "        weak_orientation = orient_list[0]\n",
    "        strong_orientation = orient_list[1]\n",
    "        \n",
    "        real_orient_list = [weak_orientation for j in range(nr_weak)] + [strong_orientation for k in range(nr_strong)] \n",
    "        \n",
    "        if len(real_orient_list) == df_len:\n",
    "            seq_coords_df[\"orientation\"] = real_orient_list\n",
    "        else:\n",
    "            raise ValueError(\"Check the numeber of weak and strong sites given. Expected sum of those two numbers is: %s\" % df_len)\n",
    "    \n",
    "    elif mode == \"for_each\":\n",
    "        rep_unit = seq_coords_df\n",
    "        orientation_ls = []\n",
    "        \n",
    "        for o in range(len(orient_list)):\n",
    "            orientation = orient_list[o]\n",
    "            orientation_ls = orientation_ls + [orientation for i in range(df_len)]\n",
    "            if len(seq_coords_df) != len(orientation_ls):\n",
    "                seq_coords_df = pd.concat([seq_coords_df, rep_unit], ignore_index=True)\n",
    "            \n",
    "        seq_coords_df[\"orientation\"] = orientation_ls\n",
    "        \n",
    "    seq_coords_df = seq_coords_df.drop([\"index\"], axis=1)\n",
    "    return seq_coords_df\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3bd466f2-cd9e-43cf-9492-7c40aabcc264",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>genomic_SCD</th>\n",
       "      <th>orientation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr12</td>\n",
       "      <td>35192359</td>\n",
       "      <td>35192378</td>\n",
       "      <td>-</td>\n",
       "      <td>53.656250</td>\n",
       "      <td>&lt;&lt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>102781112</td>\n",
       "      <td>102781131</td>\n",
       "      <td>-</td>\n",
       "      <td>53.500000</td>\n",
       "      <td>&lt;&lt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr7</td>\n",
       "      <td>139598311</td>\n",
       "      <td>139598330</td>\n",
       "      <td>+</td>\n",
       "      <td>0.058136</td>\n",
       "      <td>&lt;&lt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8</td>\n",
       "      <td>45342919</td>\n",
       "      <td>45342938</td>\n",
       "      <td>+</td>\n",
       "      <td>0.058319</td>\n",
       "      <td>&lt;&lt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr12</td>\n",
       "      <td>35192359</td>\n",
       "      <td>35192378</td>\n",
       "      <td>-</td>\n",
       "      <td>53.656250</td>\n",
       "      <td>&gt;&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr8</td>\n",
       "      <td>102781112</td>\n",
       "      <td>102781131</td>\n",
       "      <td>-</td>\n",
       "      <td>53.500000</td>\n",
       "      <td>&gt;&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr7</td>\n",
       "      <td>139598311</td>\n",
       "      <td>139598330</td>\n",
       "      <td>+</td>\n",
       "      <td>0.058136</td>\n",
       "      <td>&gt;&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr8</td>\n",
       "      <td>45342919</td>\n",
       "      <td>45342938</td>\n",
       "      <td>+</td>\n",
       "      <td>0.058319</td>\n",
       "      <td>&gt;&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr12</td>\n",
       "      <td>35192359</td>\n",
       "      <td>35192378</td>\n",
       "      <td>-</td>\n",
       "      <td>53.656250</td>\n",
       "      <td>&gt;&gt;&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr8</td>\n",
       "      <td>102781112</td>\n",
       "      <td>102781131</td>\n",
       "      <td>-</td>\n",
       "      <td>53.500000</td>\n",
       "      <td>&gt;&gt;&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr7</td>\n",
       "      <td>139598311</td>\n",
       "      <td>139598330</td>\n",
       "      <td>+</td>\n",
       "      <td>0.058136</td>\n",
       "      <td>&gt;&gt;&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr8</td>\n",
       "      <td>45342919</td>\n",
       "      <td>45342938</td>\n",
       "      <td>+</td>\n",
       "      <td>0.058319</td>\n",
       "      <td>&gt;&gt;&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr12</td>\n",
       "      <td>35192359</td>\n",
       "      <td>35192378</td>\n",
       "      <td>-</td>\n",
       "      <td>53.656250</td>\n",
       "      <td>&lt;&gt;&lt;&gt;&lt;&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr8</td>\n",
       "      <td>102781112</td>\n",
       "      <td>102781131</td>\n",
       "      <td>-</td>\n",
       "      <td>53.500000</td>\n",
       "      <td>&lt;&gt;&lt;&gt;&lt;&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr7</td>\n",
       "      <td>139598311</td>\n",
       "      <td>139598330</td>\n",
       "      <td>+</td>\n",
       "      <td>0.058136</td>\n",
       "      <td>&lt;&gt;&lt;&gt;&lt;&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr8</td>\n",
       "      <td>45342919</td>\n",
       "      <td>45342938</td>\n",
       "      <td>+</td>\n",
       "      <td>0.058319</td>\n",
       "      <td>&lt;&gt;&lt;&gt;&lt;&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom      start        end strand  genomic_SCD orientation\n",
       "0   chr12   35192359   35192378      -    53.656250          <<\n",
       "1    chr8  102781112  102781131      -    53.500000          <<\n",
       "2    chr7  139598311  139598330      +     0.058136          <<\n",
       "3    chr8   45342919   45342938      +     0.058319          <<\n",
       "4   chr12   35192359   35192378      -    53.656250          >>\n",
       "5    chr8  102781112  102781131      -    53.500000          >>\n",
       "6    chr7  139598311  139598330      +     0.058136          >>\n",
       "7    chr8   45342919   45342938      +     0.058319          >>\n",
       "8   chr12   35192359   35192378      -    53.656250         >>>\n",
       "9    chr8  102781112  102781131      -    53.500000         >>>\n",
       "10   chr7  139598311  139598330      +     0.058136         >>>\n",
       "11   chr8   45342919   45342938      +     0.058319         >>>\n",
       "12  chr12   35192359   35192378      -    53.656250      <><><>\n",
       "13   chr8  102781112  102781131      -    53.500000      <><><>\n",
       "14   chr7  139598311  139598330      +     0.058136      <><><>\n",
       "15   chr8   45342919   45342938      +     0.058319      <><><>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "add_orientation(seq_coords_df, \n",
    "                nr_strong = 2,\n",
    "                nr_weak = 2,\n",
    "                mode=\"for_each\",\n",
    "                orient_list=[\"<<\", \">>\", \">>>\", \"<><><>\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a0566f2-9a18-4b5b-99b0-12bfbc7bd024",
   "metadata": {},
   "outputs": [],
   "source": [
    "duap[0].shape\n",
    "# 1st dim - numer of paddings tested\n",
    "# 2nd dim - number of sequences\n",
    "# 3rd dim - number of bINSackground sequences\n",
    "# 4th dim - number of targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a781909-16aa-4252-a2de-2a254bec25a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "duap[1][16].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f45bfb3-4e47-476f-8b7a-6a5ae5c6461f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# seq_coords_df.to_csv(\"./50-60pads_divergent.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f787be25-e0bb-4506-b7b8-d1c96ab272de",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_h5(seq_coords_df, out_dir, filename, stat, prediction):\n",
    "    chrom = seq_coords_df.chrom.to_numpy()\n",
    "    start = seq_coords_df.start.to_numpy().astype(np.int32)\n",
    "    end = seq_coords_df.end.to_numpy().astype(np.int32)\n",
    "    strand = seq_coords_df.strand.to_numpy()\n",
    "    genSCD = seq_coords_df.genomic_SCD.to_numpy().astype(np.float64)\n",
    "    \n",
    "    with h5py.File(out_dir + \"/\" + filename, \"w\") as hf:\n",
    "        hf.create_dataset(\"SCD\",  data=prediction[0])\n",
    "        for key in prediction[1]:\n",
    "            hf.create_dataset(\"INS-\" + str(key),  data=prediction[1][key])\n",
    "        hf.create_dataset(\"Chromosome\",  data=chrom.astype('S'))\n",
    "        hf.create_dataset(\"Start\",  data=start)\n",
    "        hf.create_dataset(\"End\",  data=end)\n",
    "        hf.create_dataset(\"Strand\",  data=strand.astype('S'))\n",
    "        hf.create_dataset(\"genomic_SCD\",  data=genSCD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43a1b69c-aa5b-4c36-b030-ad846d164649",
   "metadata": {},
   "outputs": [],
   "source": [
    "save_h5(seq_coords_df=seq_coords_df, \n",
    "        out_dir=\".\", \n",
    "        filename=\"test.h5\", \n",
    "        stat=[\"SCD\", \"INS-16\", \"INS-20\"], \n",
    "        prediction=duap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f62dc58-fb57-4a4d-90d6-a5b719742422",
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py\n",
    "\n",
    "with h5py.File('test.h5', 'w') as hf:\n",
    "    hf.create_dataset(\"SCD\",  data=duap[0])\n",
    "    for key in duap[1]:\n",
    "        hf.create_dataset(\"INS-\" + str(key),  data=duap[0])\n",
    "    hf.create_dataset(\"Chromosome\",  data=chrom.astype('S'))\n",
    "    hf.create_dataset(\"Start\",  data=start)\n",
    "    hf.create_dataset(\"End\",  data=end)\n",
    "    hf.create_dataset(\"Strand\",  data=strand.astype('S'))\n",
    "    hf.create_dataset(\"genomic_SCD\",  data=genSCD)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e783107a-08b5-4a93-9102-fde4a73f2cb1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Basenji kernel",
   "language": "python",
   "name": "basenji_kernel"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
