{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5e87b1ba-6daf-42bc-9931-17981340685a",
   "metadata": {},
   "source": [
    "This notebook compares the results of genome-wide boundary CTCF perturbations with DNA shape.\n",
    "\n",
    "First, in silico perturbation results are loaded from h5 and filtered based on whether a given boundary was responsive to CTCF motif permutation.\n",
    "\n",
    "Sites are further filtered to have similar jaspar CTCF scores for the core motif.\n",
    "\n",
    "Sequences around each site are loaded and converted to DNA shape.\n",
    "\n",
    "The propeller twist (proT) is somewhat different between motifs that have strong & weak effects on SCD/insulation, but not as dramatically as when no restriction is placed on the core motif Jaspar score. However, this likely just reflects different GC content ~10bp upstream of strong motifs.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "881fe14f-e758-4108-bc95-c0cb75361b7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import bioframe\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import h5py\n",
    "import os\n",
    "from scipy.stats import spearmanr, pearsonr\n",
    "from io import StringIO\n",
    "import pysam\n",
    "\n",
    "from matplotlib import colors\n",
    "\n",
    "cmap_purple_green_GC = colors.ListedColormap(\n",
    "    [\"mistyrose\", \"darkseagreen\", \"plum\", \"beige\"]  # a  # c  # g  # t\n",
    ")\n",
    "cmap_rhee = colors.ListedColormap(\n",
    "    [\"red\", \"blue\", \"yellow\", \"green\"]  # a red  # c blue  # g yllow  # t green\n",
    ")\n",
    "\n",
    "cmaps = [cmap_purple_green_GC, cmap_rhee]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4ca6e2cc-f836-49c8-a657-bd41f79bfb3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from akita_utils.format_io import h5_to_df\n",
    "from akita_utils.tsv_gen_utils import filter_by_chrmlen, filter_boundary_ctcfs_from_h5, filter_by_overlap_num\n",
    "from akita_utils.dna_utils import (\n",
    "    dna_1hot,\n",
    "    dna_1hot_index,\n",
    "    dna_1hot_GC,\n",
    "    dna_rc,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "565b5d1b-c41d-461e-80e2-22c45206a9c2",
   "metadata": {},
   "source": [
    "### Load & filter genomic deletion SCD for boundary CTCFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "eb542ef5-2ac3-4b1e-9745-92d7a4eeab5f",
   "metadata": {},
   "outputs": [
    {
     "ename": "AxisError",
     "evalue": "axis 1 is out of bounds for array of dimension 1",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAxisError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [3]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m df \u001b[38;5;241m=\u001b[39m \u001b[43mfilter_boundary_ctcfs_from_h5\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      2\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;66;43;03m# h5_dirs=\"/project/fudenber_735/tensorflow_models/akita/v2/analysis/permute_boundaries_motifs_ctcf_mm10_model*/scd.h5\",\u001b[39;49;00m\n\u001b[1;32m      3\u001b[0m \u001b[43m    \u001b[49m\u001b[43mh5_dirs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/scratch1/kamulege/disruption_genomic_scds_v2/*/*/scd.h5\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      4\u001b[0m \u001b[43m    \u001b[49m\u001b[43mscore_key\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mSCD\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      5\u001b[0m \u001b[43m    \u001b[49m\u001b[43mthreshold_all_ctcf\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m5\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m      6\u001b[0m \u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/akita_utils/akita_utils/tsv_gen_utils.py:40\u001b[0m, in \u001b[0;36mfilter_boundary_ctcfs_from_h5\u001b[0;34m(h5_dirs, score_key, threshold_all_ctcf)\u001b[0m\n\u001b[1;32m     38\u001b[0m dfs \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m     39\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m h5_file \u001b[38;5;129;01min\u001b[39;00m glob\u001b[38;5;241m.\u001b[39mglob(h5_dirs):\n\u001b[0;32m---> 40\u001b[0m     dfs\u001b[38;5;241m.\u001b[39mappend(\u001b[43makita_utils\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mformat_io\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mh5_to_df\u001b[49m\u001b[43m(\u001b[49m\u001b[43mh5_file\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[1;32m     41\u001b[0m df \u001b[38;5;241m=\u001b[39m dfs[\u001b[38;5;241m0\u001b[39m]\u001b[38;5;241m.\u001b[39mcopy()\n\u001b[1;32m     42\u001b[0m \u001b[38;5;66;03m# print([df[score_key] for df in dfs])\u001b[39;00m\n",
      "File \u001b[0;32m~/akita_utils/akita_utils/format_io.py:46\u001b[0m, in \u001b[0;36mh5_to_df\u001b[0;34m(filename, scd_stats, drop_duplicates_key, verbose)\u001b[0m\n\u001b[1;32m     44\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m verbose:\n\u001b[1;32m     45\u001b[0m         \u001b[38;5;28mprint\u001b[39m(key)\n\u001b[0;32m---> 46\u001b[0m     s\u001b[38;5;241m.\u001b[39mappend(pd\u001b[38;5;241m.\u001b[39mSeries(\u001b[43mhf\u001b[49m\u001b[43m[\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m]\u001b[49m\u001b[43m[\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmean\u001b[49m\u001b[43m(\u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m, name\u001b[38;5;241m=\u001b[39mkey))\n\u001b[1;32m     47\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m     48\u001b[0m     s\u001b[38;5;241m.\u001b[39mappend(pd\u001b[38;5;241m.\u001b[39mSeries(hf[key][()], name\u001b[38;5;241m=\u001b[39mkey))\n",
      "File \u001b[0;32m~/.conda/envs/basenji-gpu/lib/python3.8/site-packages/numpy/core/_methods.py:167\u001b[0m, in \u001b[0;36m_mean\u001b[0;34m(a, axis, dtype, out, keepdims, where)\u001b[0m\n\u001b[1;32m    163\u001b[0m arr \u001b[38;5;241m=\u001b[39m asanyarray(a)\n\u001b[1;32m    165\u001b[0m is_float16_result \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[0;32m--> 167\u001b[0m rcount \u001b[38;5;241m=\u001b[39m \u001b[43m_count_reduce_items\u001b[49m\u001b[43m(\u001b[49m\u001b[43marr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkeepdims\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mkeepdims\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mwhere\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mwhere\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    168\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m rcount \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m where \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m \u001b[38;5;28;01melse\u001b[39;00m umr_any(rcount \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m, axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    169\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mMean of empty slice.\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;167;01mRuntimeWarning\u001b[39;00m, stacklevel\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m2\u001b[39m)\n",
      "File \u001b[0;32m~/.conda/envs/basenji-gpu/lib/python3.8/site-packages/numpy/core/_methods.py:76\u001b[0m, in \u001b[0;36m_count_reduce_items\u001b[0;34m(arr, axis, keepdims, where)\u001b[0m\n\u001b[1;32m     74\u001b[0m     items \u001b[38;5;241m=\u001b[39m nt\u001b[38;5;241m.\u001b[39mintp(\u001b[38;5;241m1\u001b[39m)\n\u001b[1;32m     75\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m ax \u001b[38;5;129;01min\u001b[39;00m axis:\n\u001b[0;32m---> 76\u001b[0m         items \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m=\u001b[39m arr\u001b[38;5;241m.\u001b[39mshape[\u001b[43mmu\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mnormalize_axis_index\u001b[49m\u001b[43m(\u001b[49m\u001b[43max\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43marr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mndim\u001b[49m\u001b[43m)\u001b[49m]\n\u001b[1;32m     77\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m     78\u001b[0m     \u001b[38;5;66;03m# TODO: Optimize case when `where` is broadcast along a non-reduction\u001b[39;00m\n\u001b[1;32m     79\u001b[0m     \u001b[38;5;66;03m# axis and full sum is more excessive than needed.\u001b[39;00m\n\u001b[1;32m     80\u001b[0m \n\u001b[1;32m     81\u001b[0m     \u001b[38;5;66;03m# guarded to protect circular imports\u001b[39;00m\n\u001b[1;32m     82\u001b[0m     \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mnumpy\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mlib\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mstride_tricks\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m broadcast_to\n",
      "\u001b[0;31mAxisError\u001b[0m: axis 1 is out of bounds for array of dimension 1"
     ]
    }
   ],
   "source": [
    "df = filter_boundary_ctcfs_from_h5(\n",
    "    # h5_dirs=\"/project/fudenber_735/tensorflow_models/akita/v2/analysis/permute_boundaries_motifs_ctcf_mm10_model*/scd.h5\",\n",
    "    h5_dirs=\"/scratch1/kamulege/disruption_genomic_scds_v2/*/*/scd.h5\",\n",
    "    score_key=\"SCD\",\n",
    "    threshold_all_ctcf=5,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7187514-4911-4a2c-b941-7bbab1bd69de",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "393bb3de-b830-42b9-ad75-7c23b8661ab2",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(10, 4))\n",
    "\n",
    "plt.subplot(121)\n",
    "plt.scatter(df[\"score_10k\"], df[\"score_all_ctcf\"], 5, alpha=0.2)\n",
    "plt.loglog()\n",
    "plt.xlabel(\"score mut 10k\")\n",
    "plt.ylabel(\"score all CTCF\")\n",
    "\n",
    "plt.subplot(122)\n",
    "plt.scatter(df[\"score_10k\"], df[\"SCD\"], 5, alpha=0.2)\n",
    "plt.loglog()\n",
    "plt.xlabel(\"score mut 10k\")\n",
    "plt.ylabel(\"SCD individual CTCF\")\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2d5e034-7ae9-46b1-833d-dc390281ad29",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter by rmsk\n",
    "rmsk_file = \"/project/fudenber_735/genomes/mm10/database/rmsk.txt.gz\"\n",
    "rmsk_cols = list(\n",
    "    pd.read_csv(\n",
    "        StringIO(\n",
    "            \"\"\"bin swScore milliDiv milliDel milliIns genoName genoStart genoEnd genoLeft strand repName repClass repFamily repStart repEnd repLeft id\"\"\"\n",
    "        ),\n",
    "        sep=\" \",\n",
    "    )\n",
    ")\n",
    "\n",
    "rmsk = pd.read_table(\n",
    "    rmsk_file,\n",
    "    names=rmsk_cols,\n",
    ")\n",
    "rmsk.rename(\n",
    "    columns={\"genoName\": \"chrom\", \"genoStart\": \"start\", \"genoEnd\": \"end\"},\n",
    "    inplace=True,\n",
    ")\n",
    "\n",
    "sites = filter_by_overlap_num(df, rmsk, working_df_cols=[\"chrom\", \"start_2\", \"end_2\"])\n",
    "\n",
    "\n",
    "# sites.head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d3b1bfb-ca31-4745-bc5d-ebcc8059e2b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7947b6be-1467-42c6-831d-52f67210e992",
   "metadata": {},
   "outputs": [],
   "source": [
    "## add scores for jaspar core motif to table of sites\n",
    "ctcf_motifs = bioframe.read_table(\n",
    "    \"/project/fudenber_735/motifs/mm10/jaspar/MA0139.1.tsv.gz\",\n",
    "    schema=\"jaspar\",\n",
    "    skiprows=1,\n",
    ")\n",
    "sites = sites.merge(\n",
    "    ctcf_motifs,\n",
    "    how=\"left\",\n",
    "    left_on=[\"chrom\", \"start_2\", \"end_2\"],\n",
    "    right_on=[\"chrom\", \"start\", \"end\"],\n",
    ")\n",
    "\n",
    "sites_unfiltered = sites.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a7fb1cd-cb63-43fb-b0fc-3018afd13f82",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(8, 4))\n",
    "plt.subplot(121)\n",
    "plt.scatter(sites[\"score\"], sites[\"SCD\"], 10, alpha=0.5)\n",
    "plt.semilogy()\n",
    "plt.xlabel(\"core motif jaspar score\")\n",
    "plt.ylabel(\"SCD deletion score\")\n",
    "\n",
    "plt.subplot(122)\n",
    "plt.scatter(sites[\"score\"], sites[\"INS-64\"], 10, alpha=0.5)\n",
    "plt.xlabel(\"core motif jaspar score\")\n",
    "plt.ylabel(\"ins-64 deletion score\")\n",
    "\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d52eb687-4311-4a59-8aaa-c9e0bedd6699",
   "metadata": {},
   "outputs": [],
   "source": [
    "## get sites with similar core motif strengths\n",
    "# sites = sites.query('(score > 800) and (score < 850)')\n",
    "\n",
    "sites = sites_unfiltered.query(\"(score > 850) and (score < 875)\").copy()\n",
    "sites.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a198c4f-5b1b-46b2-b40a-b0397bf72096",
   "metadata": {},
   "outputs": [],
   "source": [
    "## split sites by strong vs weak, here using insulation score (ins-64)\n",
    "num_sites = 100\n",
    "sites_weak = sites[\"INS-64\"].values > -0.01\n",
    "sites_strong = sites[\"INS-64\"].values < -0.03\n",
    "\n",
    "sites_weak = np.where(sites_weak)[0]\n",
    "sites_strong = np.where(sites_strong)[0]\n",
    "print(\"num weak\", len(sites_weak), \", num strong\", len(sites_strong))\n",
    "\n",
    "sites_weak = sites_weak[\n",
    "    np.sort(np.random.choice(len(sites_weak), num_sites, replace=False))\n",
    "]\n",
    "sites_strong = sites_strong[\n",
    "    np.sort(np.random.choice(len(sites_strong), num_sites, replace=False))\n",
    "]\n",
    "sites_strong = sites.loc[sites_strong].reset_index(drop=True)\n",
    "sites_weak = sites.loc[sites_weak].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c22a32d5-65e5-4fee-aa2b-bff2088cf5d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_seqs(\n",
    "    sites,\n",
    "    genome_file=\"/project/fudenber_735/genomes/mm10/mm10.fa\",\n",
    "    extent_bp=100,\n",
    "    verbose=False,\n",
    "):\n",
    "    \"\"\"\n",
    "    Get and orient DNA sequence around motifs, and return arrays with various transformation:\n",
    "    into 1hot, index, and binary (GC).\n",
    "    \"\"\"\n",
    "    genome_open = pysam.Fastafile(genome_file)\n",
    "\n",
    "    seqs = []\n",
    "    seqs_1hot = []\n",
    "    seqs_index = []\n",
    "    seqs_GC = []\n",
    "    for i in range(len(sites)):\n",
    "        if verbose:\n",
    "            if np.mod(i, 50) == 0:\n",
    "                print(i)\n",
    "        chrm, start, end, strand = sites.loc[i][\n",
    "            [\"chrom\", \"start_2\", \"end_2\", \"strand_2\"]\n",
    "        ]\n",
    "        start = start - extent_bp\n",
    "        end = end + extent_bp\n",
    "        seq = genome_open.fetch(chrm, start, end).upper()\n",
    "        if strand == \"-\":\n",
    "            seq = dna_rc(seq)\n",
    "        seqs.append(seq)\n",
    "        seqs_1hot.append(dna_1hot(seq))\n",
    "        seqs_index.append(dna_1hot_index(seq))\n",
    "        seqs_GC.append(dna_1hot_GC(seq))\n",
    "    seqs_index = np.array(seqs_index)\n",
    "    seqs_GC = np.array(seqs_GC)\n",
    "    seqs_1hot = np.array(seqs_1hot)\n",
    "    return seqs, seqs_1hot, seqs_index, seqs_GC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b72f4604-eb72-4e36-8463-0359570abf6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs_weak, seqs_1hot_weak, seqs_index_weak, seqs_GC_weak = get_seqs(\n",
    "    sites_weak, genome_file=\"/project/fudenber_735/genomes/mm10/mm10.fa\", extent_bp=100\n",
    ")\n",
    "seqs_strong, seqs_1hot_strong, seqs_index_strong, seqs_GC_strong = get_seqs(\n",
    "    sites_strong,\n",
    "    genome_file=\"/project/fudenber_735/genomes/mm10/mm10.fa\",\n",
    "    extent_bp=100,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "714f0b2a-8532-4d15-aee6-83e424909b37",
   "metadata": {},
   "outputs": [],
   "source": [
    "#### load in Rohs lab DNA shape table\n",
    "dnashape = pd.read_excel(\n",
    "    \"/project/fudenber_735/motifs/DNA shape query table.xlsx\", skipfooter=4\n",
    ")\n",
    "dnashape_dict = dnashape.set_index(\"Pentamer\")  # .loc['AAAAAA']\n",
    "dnashape_rc = dnashape.copy()\n",
    "for i in range(len(dnashape)):\n",
    "    dnashape_rc.loc[i, \"Pentamer\"] = dna_rc(dnashape_rc.loc[i, \"Pentamer\"])\n",
    "dnashape_dict = pd.concat([dnashape, dnashape_rc])\n",
    "dnashape_dict = dnashape_dict.set_index(\"Pentamer\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "508a9866-ee1d-4ca5-91a7-eff4a31a2928",
   "metadata": {},
   "outputs": [],
   "source": [
    "PENTAMER_LEN = 5\n",
    "\n",
    "\n",
    "def shape_converter(seq, shape_key=\"MGW\"):\n",
    "    shape_vec = []\n",
    "    for i in range(len(seq) - PENTAMER_LEN):\n",
    "        subseq = seq[i : i + PENTAMER_LEN]\n",
    "        if \"N\" in subseq:\n",
    "            shape_vec.append(np.nan)\n",
    "        else:\n",
    "            shape_vec.append(dnashape_dict[shape_key].loc[subseq])\n",
    "    return np.array(shape_vec)\n",
    "\n",
    "\n",
    "def shape_converter_multi(seq_list, shape_key=\"MGW\"):\n",
    "    shape_mat = []\n",
    "    for seq in seq_list:\n",
    "        shape_mat.append(shape_converter(seq, shape_key))\n",
    "    return np.array(shape_mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa3c7434-42a4-4d59-a057-e967e7f71d01",
   "metadata": {},
   "outputs": [],
   "source": [
    "shape_val_dict = {}\n",
    "for shape_key in dnashape.keys()[1:]:\n",
    "    shape_weak = shape_converter_multi(seqs_weak, shape_key=shape_key)\n",
    "    shape_strong = shape_converter_multi(seqs_strong, shape_key=shape_key)\n",
    "    shape_val_dict[shape_key] = [shape_weak, shape_strong]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "390748b7-3202-4bd5-a55d-ddd1de4f7f3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "shape_corrs = {}\n",
    "bp_start, bp_stop = 80, 139\n",
    "inds_no_center = np.hstack((np.arange(20), np.arange(39, bp_stop - bp_start)))\n",
    "\n",
    "for shape_key in dnashape.keys()[1:]:\n",
    "    weak_shape, strong_shape = shape_val_dict[shape_key]\n",
    "\n",
    "    shape_corrs[shape_key] = [\n",
    "        np.round(\n",
    "            spearmanr(\n",
    "                np.nanmean(weak_shape[:, bp_start:bp_stop], axis=0)[inds_no_center],\n",
    "                np.nanmean(strong_shape[:, bp_start:bp_stop], axis=0)[inds_no_center],\n",
    "            )[0],\n",
    "            2,\n",
    "        ),\n",
    "        np.round(\n",
    "            np.sum(\n",
    "                np.abs(\n",
    "                    np.nanmean(weak_shape[:, bp_start:bp_stop], axis=0)[inds_no_center]\n",
    "                    - np.nanmean(strong_shape[:, bp_start:bp_stop], axis=0)[\n",
    "                        inds_no_center\n",
    "                    ]\n",
    "                )\n",
    "            ),\n",
    "            2,\n",
    "        ),\n",
    "    ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a870df97-7e8e-400f-960a-0bc6cb9374b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(5, 5))\n",
    "for k, v in shape_corrs.items():\n",
    "    plt.plot(v[0], v[1], \"o\")\n",
    "    plt.text(v[0], v[1], k)\n",
    "plt.grid()\n",
    "plt.xlabel(\"corr\")\n",
    "plt.ylabel(\"total diff\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24d2e03d-e922-45ab-ac4c-46f2880ff8c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "for shape_key in dnashape.keys()[1:]:\n",
    "    weak_shape, strong_shape = shape_val_dict[shape_key]\n",
    "\n",
    "    min_val, max_val = np.nanpercentile(\n",
    "        # [weak_shape , strong_shape ],[5,95])\n",
    "        [weak_shape, strong_shape],\n",
    "        [10, 90],\n",
    "    )\n",
    "    print(shape_key, min_val, max_val)\n",
    "\n",
    "    bp_start, bp_stop = (80, 139)\n",
    "\n",
    "    plt.figure(figsize=(10, 6))\n",
    "    plt.subplot(121)\n",
    "    im = plt.matshow(\n",
    "        weak_shape[:, bp_start:bp_stop],\n",
    "        vmin=min_val,\n",
    "        vmax=max_val,\n",
    "        cmap=\"jet\",\n",
    "        fignum=False,\n",
    "    )\n",
    "    plt.colorbar(im, fraction=0.046, pad=0.04)\n",
    "    plt.title(\"weak \" + str(shape_key))\n",
    "    ax = plt.gca()\n",
    "    ax.set_aspect(\"auto\")\n",
    "\n",
    "    plt.subplot(122)\n",
    "    im = plt.matshow(\n",
    "        strong_shape[:, bp_start:bp_stop],\n",
    "        vmin=min_val,\n",
    "        vmax=max_val,\n",
    "        cmap=\"jet\",\n",
    "        fignum=False,\n",
    "    )\n",
    "    plt.colorbar(im, fraction=0.046, pad=0.04)\n",
    "    plt.title(\"strong \" + str(shape_key))\n",
    "    # plt.tight_layout()\n",
    "    ax.set_aspect(\"auto\")\n",
    "    ax = plt.gca()\n",
    "    ax.set_aspect(\"auto\")\n",
    "    plt.show()\n",
    "\n",
    "    plt.figure(figsize=(9.5, 1))\n",
    "    plt.subplot(121)\n",
    "    plt.plot(np.mean(weak_shape[:, bp_start:bp_stop], axis=0))\n",
    "    plt.grid()\n",
    "    plt.ylim([min_val, max_val])\n",
    "    plt.subplot(122)\n",
    "    plt.plot(np.mean(weak_shape[:, bp_start:bp_stop], axis=0))\n",
    "\n",
    "    plt.plot(np.mean(strong_shape[:, bp_start:bp_stop], axis=0), label=\"strong\")\n",
    "    plt.grid()\n",
    "    plt.ylim([min_val, max_val])\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9144ccdb-031d-4afe-96ca-397ec0454738",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "basenji-gpu",
   "language": "python",
   "name": "basenji-gpu"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
