{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8154d563-af62-473c-a8f4-26f482ad9e75",
   "metadata": {},
   "source": [
    "# In this notebook we explore the influnce of different ways of masking CTCF motis to the prediction score and success time.\n",
    "\n",
    "There are three ways to go about this (in brackets are corresponding tags);\n",
    "- just randomly shuffle the whole seq i.e no masking (all random)\n",
    "- scan for CTCF and randomly shuffle the motifs (random motif masking)\n",
    "- scan for CTCF and manually shuffle the motifs (manual motif masking)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "881fe14f-e758-4108-bc95-c0cb75361b7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "import bioframe \n",
    "import pandas as pd\n",
    "import numpy as np \n",
    "import os\n",
    "import json\n",
    "from io import StringIO\n",
    "import seaborn as sns\n",
    "import random \n",
    "import akita_utils \n",
    "import pysam\n",
    "import time \n",
    "import h5py\n",
    "genome_fasta = '/project/fudenber_735/genomes/mm10/mm10.fa'\n",
    "genome_open = pysam.Fastafile('/project/fudenber_735/genomes/mm10/mm10.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8951f01a-2fed-4230-ba35-e8f1469a2e69",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package akita_utils:\n",
      "\n",
      "NAME\n",
      "    akita_utils\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    __main__\n",
      "    background_utils\n",
      "    base\n",
      "    cli\n",
      "    slurm_gf\n",
      "    utils\n",
      "\n",
      "FILE\n",
      "    /home1/kamulege/akita_utils/akita_utils/__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(akita_utils)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "40a0bf0d-860a-4b3c-9ef9-13dd4f319b7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.4.1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-11-29 21:25:28.765826: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2 AVX AVX2 FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "built\n",
      "restored\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(512, 130305, (130305,))"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "os.environ[\"CUDA_VISIBLE_DEVICES\"] = '-1'\n",
    "import tensorflow as tf\n",
    "print(tf.__version__)\n",
    "\n",
    "import pysam\n",
    "from basenji import dataset, seqnn, dna_io,stream\n",
    "head_i = 1 #mouse\n",
    "model_num = 1 #which fold to use\n",
    "\n",
    "base_dir = '/project/fudenber_735/tensorflow_models/akita/v2/models/'\n",
    "model_dir = base_dir+\"/f\"+str(model_num)+\"c0/train/\"\n",
    "model_file  = model_dir+'/model'+str(head_i)+'_best.h5'\n",
    "\n",
    "params_file = model_dir+'/params.json'\n",
    "with open(params_file) as params_open:\n",
    "    params = json.load(params_open)\n",
    "    params_model = params['model']\n",
    "    params_train = params['train']\n",
    "seq_length = params_model['seq_length']\n",
    "params_model['verbose'] = False\n",
    "seqnn_model = seqnn.SeqNN(params_model)\n",
    "print('built')\n",
    "seqnn_model.restore(model_file, head_i=head_i)\n",
    "print('restored')\n",
    "\n",
    "hic_diags = params_model['diagonal_offset']\n",
    "try:\n",
    "    target_crop = params_model['trunk'][-2]['cropping']\n",
    "except:\n",
    "    target_crop = params_model['target_crop']\n",
    "target_length_cropped = int((seq_length//2048 - target_crop*2 - hic_diags)* ((seq_length//2048 - target_crop*2 - hic_diags) +1)/2) \n",
    "target_map_size = seq_length//2048  - target_crop*2 \n",
    "triu_tup = np.triu_indices(target_map_size,2)\n",
    "target_map_size, target_length_cropped, triu_tup[0].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "813b2086-efa8-41c7-a5b3-5e1ae2728031",
   "metadata": {},
   "source": [
    "# Fetching chromosome data to explore background creation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9c6cd103-e89a-4963-9fa4-e8d24e2b54ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>GC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>40415232</td>\n",
       "      <td>41725952</td>\n",
       "      <td>0.407777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42708992</td>\n",
       "      <td>44019712</td>\n",
       "      <td>0.424985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>42053632</td>\n",
       "      <td>43364352</td>\n",
       "      <td>0.424847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>41725952</td>\n",
       "      <td>43036672</td>\n",
       "      <td>0.411407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>43036672</td>\n",
       "      <td>44347392</td>\n",
       "      <td>0.417725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6033</th>\n",
       "      <td>chrX</td>\n",
       "      <td>99241984</td>\n",
       "      <td>100552704</td>\n",
       "      <td>0.424162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6034</th>\n",
       "      <td>chrX</td>\n",
       "      <td>120532992</td>\n",
       "      <td>121843712</td>\n",
       "      <td>0.367963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6035</th>\n",
       "      <td>chrX</td>\n",
       "      <td>139360256</td>\n",
       "      <td>140670976</td>\n",
       "      <td>0.407839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6036</th>\n",
       "      <td>chrX</td>\n",
       "      <td>128737280</td>\n",
       "      <td>130048000</td>\n",
       "      <td>0.370827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6037</th>\n",
       "      <td>chrX</td>\n",
       "      <td>101535744</td>\n",
       "      <td>102846464</td>\n",
       "      <td>0.430301</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6038 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end        GC\n",
       "0     chr1   40415232   41725952  0.407777\n",
       "1     chr1   42708992   44019712  0.424985\n",
       "2     chr1   42053632   43364352  0.424847\n",
       "3     chr1   41725952   43036672  0.411407\n",
       "4     chr1   43036672   44347392  0.417725\n",
       "...    ...        ...        ...       ...\n",
       "6033  chrX   99241984  100552704  0.424162\n",
       "6034  chrX  120532992  121843712  0.367963\n",
       "6035  chrX  139360256  140670976  0.407839\n",
       "6036  chrX  128737280  130048000  0.370827\n",
       "6037  chrX  101535744  102846464  0.430301\n",
       "\n",
       "[6038 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chromsizes = bioframe.read_chromsizes('/project/fudenber_735/tensorflow_models/akita/v2/data/mm10/sequences.bed')\n",
    "dframe = pd.DataFrame(chromsizes)\n",
    "dframe['end'] = dframe['length']+ 1310720\n",
    "dframe = dframe.reset_index()\n",
    "dframe.rename(columns = {'index' : 'chrom', 'length':'start'}, inplace = True)\n",
    "df = bioframe.frac_gc(dframe, bioframe.load_fasta('/project/fudenber_735/genomes/mm10/mm10.fa'), return_input=True)\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "65002b84-7a13-455e-b290-a893f4246508",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df.iloc[6][['chrom','start','end','GC']]\n",
    "# plt.hist( df['GC'].values, 100);\n",
    "# np.percentile(df['GC'].dropna().values, np.linspace(1,99,50))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e66a7411-5169-4011-aeb7-0a8067155ff9",
   "metadata": {},
   "source": [
    "# Generating a sample for down stream analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e4cf45c4-4b2d-4c1e-8b88-abcef8c719c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Whole distribution: [0, 1, 3, 4, 6, 7, 8, 9, 10, 12, 13, 15, 16, 147, 19, 21, 22, 23, 24, 26, 28, 288, 33, 163, 35, 165, 38, 39, 171, 44, 45, 47, 303, 50, 567, 64, 65, 68, 329, 74, 77, 79, 85, 90, 102, 496, 119, 120, 122] \n",
      "\n",
      "Sampled datapoints: [3, 4, 6, 7, 8, 9, 10, 12]\n"
     ]
    }
   ],
   "source": [
    "super_set = []\n",
    "error = 0.001\n",
    "\n",
    "for gc in np.percentile(df['GC'].dropna().values, np.linspace(1,99,50)):\n",
    "    for i in range(df.shape[0]):\n",
    "        if gc-error <= df['GC'].values[i] <= gc+error:\n",
    "            super_set += [i]\n",
    "            break\n",
    "\n",
    "super_set = list(set(super_set)); print(f'Whole distribution: {super_set} \\n')\n",
    "sample_set = super_set[2:10]; print(f'Sampled datapoints: {sample_set}')\n",
    "new_dataframe = df.iloc[[ind for ind in set(sample_set)]] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "95109e96-4d99-4542-ac06-aead38c14700",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutation_search(seqnn_model,\n",
    "                    genome_fasta,\n",
    "                    seq_length,\n",
    "                    dataframe,\n",
    "                    max_iters=1,\n",
    "                    batch_size=6,\n",
    "                    shuffle_k=8,\n",
    "                    ctcf_thresh=8,\n",
    "                    scores_thresh = 5500,\n",
    "                    scores_pixelwise_thresh = .04,\n",
    "                    masking=False,\n",
    "                    timing = False\n",
    "                    ):\n",
    "    \"\"\"\n",
    "    This function creates flat sequences\n",
    "    \"\"\"\n",
    "    flat_seqs = []\n",
    "    scores_storage = {}\n",
    "    num_seqs = dataframe.shape[0]\n",
    "\n",
    "    mot = \">CCAsyAGrkGGCr\\n0.0000\\t1.0000\\t0.0000\\t0.0000\\n0.0000\\t1.0000\\t0.0000\\t0.0000\\n1.0000\\t0.0000\\t0.0000\\t0.0000\\n0.0000\\t0.5000\\t0.5000\\t0.0000\\n0.0000\\t0.5000\\t0.0000\\t0.5000\\n1.0000\\t0.0000\\t0.0000\\t0.0000\\n0.0000\\t0.0000\\t1.0000\\t0.0000\\n0.5000\\t0.0000\\t0.5000\\t0.0000\\n0.0000\\t0.0000\\t0.5000\\t0.5000\\n0.0000\\t0.0000\\t1.0000\\t0.0000\\n0.0000\\t0.0000\\t1.0000\\t0.0000\\n0.0000\\t1.0000\\t0.0000\\t0.0000\\n0.5000\\t0.0000\\t0.5000\\t0.0000\"\n",
    "    motif = pd.read_csv(StringIO(mot), sep=\"\\t\", header=0, names=[\"A\", \"C\", \"G\", \"T\"]).values\n",
    "    motif_window = int(np.ceil(len(motif) / 2))\n",
    "    mot_shuf = [12, 0, 1, 11, 10, 3, 2, 8, 9, 4, 5, 7, 6, 13]\n",
    "\n",
    "    for ind in range(num_seqs):\n",
    "        chrom, start, end, gc = dataframe.iloc[ind][[\"chrom\", \"start\", \"end\", \"GC\"]]\n",
    "        genome_open = pysam.Fastafile(genome_fasta)\n",
    "        seq = genome_open.fetch(chrom, start, end).upper()\n",
    "        seq_1hot = dna_io.dna_1hot(seq)\n",
    "\n",
    "        t0 = time.time()\n",
    "        num_iters = 0\n",
    "        while num_iters < max_iters:\n",
    "            name = f'{num_iters}'\n",
    "            scores_set = []\n",
    "            print(\"ind\",ind,\", iter \",num_iters,\", k \",shuffle_k,\", for\", chrom, start, end,)\n",
    "            seq_1hot_batch = []\n",
    "            for i in range(batch_size):\n",
    "                seq_1hot_mut = akita_utils.permute_seq_k(seq_1hot, k=shuffle_k)\n",
    "                s = akita_utils.scan_motif(seq_1hot_mut, motif)\n",
    "                if masking == 0:\n",
    "                    for i in np.where(s > ctcf_thresh)[0]:\n",
    "                        if len(seq_1hot_mut[i-motif_window:i+motif_window]) == len(mot_shuf):\n",
    "                            seq_1hot_mut[i-motif_window:i+motif_window] = akita_utils.permute_seq_k(seq_1hot_mut[i-motif_window:i+motif_window], k=2)\n",
    "                    seq_1hot_batch.append(seq_1hot_mut)\n",
    "                elif masking == 1:\n",
    "                    for i in np.where(s > ctcf_thresh)[0]:\n",
    "                        if len(seq_1hot_mut[i-motif_window:i+motif_window]) == len(mot_shuf):\n",
    "                            seq_1hot_mut[i-motif_window:i+motif_window] = seq_1hot_mut[i-motif_window:i+motif_window][mot_shuf]\n",
    "\n",
    "                    seq_1hot_batch.append(seq_1hot_mut)\n",
    "                else:\n",
    "                    seq_1hot_batch.append(seq_1hot_mut)\n",
    "\n",
    "            seq_1hot_batch = np.array(seq_1hot_batch)\n",
    "            pred = seqnn_model.predict(seq_1hot_batch, batch_size=batch_size)\n",
    "            scores = np.sum(pred**2, axis=-1).sum(axis=-1)\n",
    "            scores_pixelwise = np.max(pred**2, axis=-1).max(axis=-1)  \n",
    "            t1 = time.time()\n",
    "\n",
    "            if np.any([\n",
    "                (np.min(scores) < scores_thresh)\n",
    "                , (np.min(scores_pixelwise) < scores_pixelwise_thresh)]\n",
    "            ):\n",
    "                scores_set += [scores]\n",
    "                best_ind = np.argmin(scores_pixelwise)\n",
    "                best_seq = seq_1hot_batch[best_ind]\n",
    "                best_pred = pred[best_ind]\n",
    "                best_score, best_score_pixelwise = (\n",
    "                    scores[best_ind],\n",
    "                    scores_pixelwise[best_ind],\n",
    "                )\n",
    "                \n",
    "                print(\n",
    "                    \"success: best seq, thresh\",\n",
    "                    np.min(scores),\n",
    "                    \" pixelwise\",\n",
    "                    np.min(scores_pixelwise),\n",
    "                    \"time\",\n",
    "                    t1 - t0\n",
    "                )\n",
    "                if timing == True:\n",
    "                    num_iters = max_iters\n",
    "            else:\n",
    "                scores_set += [scores]\n",
    "                best_ind = np.argmin(scores_pixelwise)\n",
    "                best_seq = seq_1hot_batch[best_ind]\n",
    "                best_pred = pred[best_ind]\n",
    "                best_score, best_score_pixelwise = (\n",
    "                    scores[best_ind],\n",
    "                    scores_pixelwise[best_ind],\n",
    "                )\n",
    "                print(\n",
    "                \"trying: best seq, thresh\",\n",
    "                np.min(scores),\n",
    "                \" pixelwise\",\n",
    "                np.min(scores_pixelwise),\n",
    "                )\n",
    "            \n",
    "            scores_storage[name] = scores_set\n",
    "            \n",
    "            num_iters += 1\n",
    "            if num_iters >= max_iters:\n",
    "                # scores_set += [scores]\n",
    "                print(f\"max iters exceeded, final time {t1 - t0}\")\n",
    "                if gc :\n",
    "                    flat_seqs.append([\n",
    "                    best_seq,\n",
    "                    best_pred,\n",
    "                    best_score,\n",
    "                    best_score_pixelwise,\n",
    "                    t1 - t0,\n",
    "                    gc\n",
    "                ])\n",
    "                else:\n",
    "                    flat_seqs.append([\n",
    "                    best_seq,\n",
    "                    best_pred,\n",
    "                    best_score,\n",
    "                    best_score_pixelwise,\n",
    "                    t1 - t0\n",
    "                ])\n",
    "\n",
    "            \n",
    "\n",
    "    return scores_storage, flat_seqs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0460cd8-b589-47eb-b053-805c5e13d78c",
   "metadata": {},
   "source": [
    "# Focusing on the scores' distributions and time taken to obtain a success trial (with and without masking)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2be90373-955b-45b7-8a43-43d8108a8656",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ind 0 , iter  0 , k  8 , for chr1 41725952 43036672\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-11-29 21:26:28.258961: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:116] None of the MLIR optimization passes are enabled (registered 2)\n",
      "2022-11-29 21:26:28.259461: I tensorflow/core/platform/profile_utils/cpu_utils.cc:112] CPU Frequency: 2595005000 Hz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "success: best seq, thresh 4270.0  pixelwise 0.03168 time 6.580358505249023\n",
      "max iters exceeded, final time 6.580358505249023\n",
      "ind 1 , iter  0 , k  8 , for chr1 43036672 44347392\n",
      "trying: best seq, thresh 5964.0  pixelwise 0.05148\n",
      "ind 1 , iter  1 , k  8 , for chr1 43036672 44347392\n",
      "success: best seq, thresh 6160.0  pixelwise 0.03458 time 11.507014274597168\n",
      "max iters exceeded, final time 11.507014274597168\n",
      "ind 2 , iter  0 , k  8 , for chr1 39432192 40742912\n",
      "trying: best seq, thresh 7420.0  pixelwise 0.05984\n",
      "ind 2 , iter  1 , k  8 , for chr1 39432192 40742912\n",
      "trying: best seq, thresh 7452.0  pixelwise 0.04526\n",
      "ind 2 , iter  2 , k  8 , for chr1 39432192 40742912\n",
      "trying: best seq, thresh 6470.0  pixelwise 0.04828\n",
      "ind 2 , iter  3 , k  8 , for chr1 39432192 40742912\n",
      "trying: best seq, thresh 8456.0  pixelwise 0.05466\n",
      "ind 2 , iter  4 , k  8 , for chr1 39432192 40742912\n",
      "trying: best seq, thresh 9540.0  pixelwise 0.0593\n",
      "ind 2 , iter  5 , k  8 , for chr1 39432192 40742912\n",
      "trying: best seq, thresh 7468.0  pixelwise 0.0479\n",
      "ind 2 , iter  6 , k  8 , for chr1 39432192 40742912\n",
      "trying: best seq, thresh 8450.0  pixelwise 0.07007\n",
      "ind 2 , iter  7 , k  8 , for chr1 39432192 40742912\n",
      "trying: best seq, thresh 8068.0  pixelwise 0.04996\n",
      "ind 2 , iter  8 , k  8 , for chr1 39432192 40742912\n",
      "trying: best seq, thresh 7596.0  pixelwise 0.05334\n",
      "ind 2 , iter  9 , k  8 , for chr1 39432192 40742912\n",
      "trying: best seq, thresh 9020.0  pixelwise 0.06494\n",
      "max iters exceeded, final time 58.934754848480225\n",
      "ind 3 , iter  0 , k  8 , for chr1 38449152 39759872\n",
      "trying: best seq, thresh 7780.0  pixelwise 0.04855\n",
      "ind 3 , iter  1 , k  8 , for chr1 38449152 39759872\n",
      "trying: best seq, thresh 7650.0  pixelwise 0.0551\n",
      "ind 3 , iter  2 , k  8 , for chr1 38449152 39759872\n",
      "trying: best seq, thresh 7540.0  pixelwise 0.0551\n",
      "ind 3 , iter  3 , k  8 , for chr1 38449152 39759872\n",
      "trying: best seq, thresh 6940.0  pixelwise 0.05057\n",
      "ind 3 , iter  4 , k  8 , for chr1 38449152 39759872\n",
      "trying: best seq, thresh 8350.0  pixelwise 0.05057\n",
      "ind 3 , iter  5 , k  8 , for chr1 38449152 39759872\n",
      "trying: best seq, thresh 7336.0  pixelwise 0.0569\n",
      "ind 3 , iter  6 , k  8 , for chr1 38449152 39759872\n",
      "trying: best seq, thresh 7452.0  pixelwise 0.0569\n",
      "ind 3 , iter  7 , k  8 , for chr1 38449152 39759872\n",
      "trying: best seq, thresh 8390.0  pixelwise 0.05954\n",
      "ind 3 , iter  8 , k  8 , for chr1 38449152 39759872\n",
      "trying: best seq, thresh 6772.0  pixelwise 0.0681\n",
      "ind 3 , iter  9 , k  8 , for chr1 38449152 39759872\n",
      "trying: best seq, thresh 7450.0  pixelwise 0.05414\n",
      "max iters exceeded, final time 58.805742025375366\n",
      "ind 4 , iter  0 , k  8 , for chr1 38776832 40087552\n",
      "trying: best seq, thresh 8900.0  pixelwise 0.06287\n",
      "ind 4 , iter  1 , k  8 , for chr1 38776832 40087552\n",
      "trying: best seq, thresh 8700.0  pixelwise 0.06445\n",
      "ind 4 , iter  2 , k  8 , for chr1 38776832 40087552\n"
     ]
    }
   ],
   "source": [
    "max_iters=10\n",
    "batch_size=6\n",
    "# score = 5500\n",
    "\n",
    "scores_storage_random_masking, flat_seqs_random_masking = mutation_search(    seqnn_model=seqnn_model,\n",
    "                                                    genome_fasta=genome_fasta,\n",
    "                                                    seq_length=seq_length,\n",
    "                                                    dataframe= new_dataframe,\n",
    "                                                    max_iters=max_iters,\n",
    "                                                    batch_size=batch_size,\n",
    "                                                    masking=0,timing = True)\n",
    "\n",
    "print('**********************************')\n",
    "\n",
    "scores_storage_motif_masking, flat_seqs_motif_masking = mutation_search(     seqnn_model=seqnn_model,\n",
    "                                                    genome_fasta=genome_fasta,\n",
    "                                                    seq_length=seq_length,\n",
    "                                                    dataframe= new_dataframe,\n",
    "                                                    max_iters=max_iters,\n",
    "                                                    batch_size=batch_size,\n",
    "                                                    masking=1,timing = True)\n",
    "\n",
    "print('**********************************')\n",
    "\n",
    "scores_storage_all_random, flat_seqs_no_mask = mutation_search(        seqnn_model=seqnn_model,\n",
    "                                                    genome_fasta=genome_fasta,\n",
    "                                                    seq_length=seq_length,\n",
    "                                                    dataframe= new_dataframe,\n",
    "                                                    max_iters=max_iters,\n",
    "                                                    batch_size=batch_size,timing = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2a069e5-365b-42f9-bbb0-af7a89018963",
   "metadata": {},
   "source": [
    "# Visualizing scores' distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88df25aa-355b-475a-ada9-f697fd5d8d99",
   "metadata": {},
   "outputs": [],
   "source": [
    "Y_1 = []\n",
    "Y_2 = []\n",
    "Y_3 = []\n",
    "\n",
    "for i in scores_storage_random_masking:\n",
    "    for j in scores_storage_random_masking[i]:\n",
    "        Y_1 += j.tolist()\n",
    "\n",
    "for i in scores_storage_motif_masking:\n",
    "    for j in scores_storage_motif_masking[i]:\n",
    "        Y_2 += j.tolist()        \n",
    "\n",
    "for i in scores_storage_all_random:\n",
    "    for j in scores_storage_all_random[i]:\n",
    "        Y_3 += j.tolist()\n",
    "        \n",
    "print(scores_storage_motif_masking)\n",
    "\n",
    "kde_1 = pd.DataFrame(Y_1, columns=[\"score\"])\n",
    "kde_2 = pd.DataFrame(Y_2, columns=[\"score\"])\n",
    "kde_3 = pd.DataFrame(Y_3, columns=[\"score\"])\n",
    "sns.kdeplot(data=kde_1, x=\"score\", bw_adjust=.2, label='random motif masking', fill=True)  \n",
    "sns.kdeplot(data=kde_2, x=\"score\", bw_adjust=.2, label='manual motif masking', fill=True)    \n",
    "sns.kdeplot(data=kde_3, x=\"score\", bw_adjust=.2, label='all random', fill=True)    \n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1a0d54d-e103-4e3e-ba3b-8af63850aedc",
   "metadata": {},
   "source": [
    "# Visualizing time taken to obtain a success trial with masking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b3237ab-358e-4999-8b7f-d5d103cd8d9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "flat_pred_time_random_masking = []\n",
    "flat_pred_time_motif_masking = []\n",
    "flat_pred_time_no_mask = []\n",
    "\n",
    "for i in range(len(sample_set)):\n",
    "    flat_pred_time_no_mask += [flat_seqs_no_mask[i][4]]\n",
    "    flat_pred_time_random_masking += [flat_seqs_random_masking[i][4]]\n",
    "    flat_pred_time_motif_masking += [flat_seqs_motif_masking[i][4]]\n",
    "\n",
    "kde = pd.DataFrame(flat_pred_time_no_mask, columns=[\"success time (s)\"])    \n",
    "sns.kdeplot(data=kde, x=\"success time (s)\", bw_adjust=.2, label='all random', fill=True)    \n",
    "\n",
    "kde = pd.DataFrame(flat_pred_time_motif_masking, columns=[\"success time (s)\"])    \n",
    "sns.kdeplot(data=kde, x=\"success time (s)\", bw_adjust=.2, label='manual motif mask', fill=True)  \n",
    "\n",
    "kde = pd.DataFrame(flat_pred_time_random_masking, columns=[\"success time (s)\"])    \n",
    "sns.kdeplot(data=kde, x=\"success time (s)\", bw_adjust=.2, label='random motif mask', fill=True)  \n",
    "\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "basenji-gpu",
   "language": "python",
   "name": "basenji-gpu"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
